clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -18584,6 +18584,29 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
         CGM.getHLSLRuntime().getLengthIntrinsic(), ArrayRef<Value *>{X},
         nullptr, "hlsl.length");
   }
+  case Builtin::BI__builtin_hlsl_normalize: {
+    Value *X = EmitScalarExpr(E->getArg(0));
+
+    assert(E->getArg(0)->getType()->hasFloatingRepresentation() &&
+           "normalize operand must have a float representation");
+
+    // scalar inputs should expect a scalar return type
+    if (!E->getArg(0)->getType()->isVectorType())
+      return Builder.CreateIntrinsic(
+          /*ReturnType=*/X->getType()->getScalarType(),
+          CGM.getHLSLRuntime().getNormalizeIntrinsic(), ArrayRef<Value *>{X},
+          nullptr, "hlsl.normalize");
+
+    // construct a vector return type for vector inputs
```

Comment:
Why do you need to do this block?

Reply:
I can't just use XVecTy directly because getAs returns a const pointer. I need to supply a non-const return type to the createIntrinsic function, but can't use anything directly from E as far as I can tell.

Reply:
I mean it shouldn't be needed to getAs<VectorType>` ret type will be the same as `X`'s type so just do `/*ReturnType=*/X->getType()` that should let you delete 18601 to 18605